ionoscloud/firewall_service.go

```diff
@@ -0,0 +1,128 @@
+		return apiResponse, fmt.Errorf("an error occured while waiting for state change dcId: %s server_id: %s nic_id %s ID: %s Response: %w", datacenterId, serverId, nicId, firewallId, errState)
+	}
+	return apiResponse, nil
+}
+
+func (fs *FirewallService) createFirewallRule(ctx context.Context, datacenterId, serverId, nicId string, firewallrule ionoscloud.FirewallRule) (*ionoscloud.FirewallRule, *ionoscloud.APIResponse, error) {
+	firewall, apiResponse, err := fs.client.FirewallRulesApi.DatacentersServersNicsFirewallrulesPost(ctx, datacenterId, serverId, nicId).Firewallrule(firewallrule).Execute()
+	apiResponse.LogInfo()
+	if err != nil {
+		return nil, apiResponse, fmt.Errorf("an error occured while creating firewall rule for dcId: %s server_id: %s nic_id ID: %s Response: (%w)", datacenterId, serverId, nicId, err)
+	}
+	// Wait, catching any errors
+	_, errState := getStateChangeConf(fs.meta, fs.schemaData, apiResponse.Header.Get("Location"), schema.TimeoutCreate).WaitForStateContext(ctx)
+	if errState != nil {
+		return nil, apiResponse, fmt.Errorf("an error occured while waiting for state change dcId: %s server_id: %s nic_id ID: %s Response: (%w)", datacenterId, serverId, nicId, errState)
+	}
+	return &firewall, apiResponse, nil
+}
+
+func SetFirewallProperties(firewall ionoscloud.FirewallRule) map[string]interface{} {
+
+	fw := map[string]interface{}{}
+	/*
+		"protocol": *firewall.Properties.Protocol,
+		"name":     *firewall.Properties.Name,
+	*/
+	if firewall.Properties != nil {
+		utils.SetPropWithNilCheck(fw, "protocol", firewall.Properties.Protocol)
+		utils.SetPropWithNilCheck(fw, "name", firewall.Properties.Name)
+		utils.SetPropWithNilCheck(fw, "source_mac", firewall.Properties.SourceMac)
+		utils.SetPropWithNilCheck(fw, "source_ip", firewall.Properties.SourceIp)
+		utils.SetPropWithNilCheck(fw, "target_ip", firewall.Properties.TargetIp)
+		utils.SetPropWithNilCheck(fw, "port_range_start", firewall.Properties.PortRangeStart)
+		utils.SetPropWithNilCheck(fw, "port_range_end", firewall.Properties.PortRangeEnd)
+		utils.SetPropWithNilCheck(fw, "type", firewall.Properties.Type)
+		if firewall.Properties.IcmpType != nil {
+			fw["icmp_type"] = strconv.Itoa(int(*firewall.Properties.IcmpType))
+		}
+		if firewall.Properties.IcmpCode != nil {
+			fw["icmp_code"] = strconv.Itoa(int(*firewall.Properties.IcmpCode))
+		}
+	}
+	return fw
+}
+
+func GetChangesInFirewallRules(oldValues, newValues []interface{}) ([]ionoscloud.FirewallruleProperties, []ionoscloud.FirewallruleProperties, error) {
+	onlyOld := utils.DifferenceSlices(oldValues, newValues)
+	onlyNew := utils.DifferenceSlices(newValues, oldValues)
+	oldFwSlice := make([]ionoscloud.FirewallruleProperties, len(onlyOld))
+	newFwSlice := make([]ionoscloud.FirewallruleProperties, len(onlyNew))
+	err := utils.DecodeInterfaceToStruct(onlyNew, newFwSlice)
+	if err != nil {
+		return nil, nil, fmt.Errorf("could not decode from %s to new values of FirewallRules %w", newValues, err)
```

Comment:
@cristiGuranIonos I think we should write `firewall rules` in all these error messages.